LoRaWAN: Add Clock Synchronization Service

martinjaeger opened the pull request:
This PR adds an implementation of the Application Layer Clock Synchronization ([TS003](https://resources.lora-alliance.org/technical-specifications/ts003-2-0-0-application-layer-clock-synchronization)) service specified by the LoRa Alliance.

Clock synchronization is required for firmware-upgrade over the air (FUOTA). This PR splits out the first parts of the entire FUOTA PR #49284 to make review easier.

As for the rest of the LoRaWAN implementation in Zephyr, it uses some functions provided by the [LoRaMac-node module](https://github.com/zephyrproject-rtos/loramac-node) developed by Semtech.

The clock synchronization process is started by the application and afterwards runs in the background in its own work queue thread.

**Testing**

The implementation has been successfully tested with [ChirpStack](https://www.chirpstack.io/) and AWS IoT.

For testing with ChirpStack, the [FUOTA server](https://github.com/brocaar/chirpstack-fuota-server) is needed in addition to the normal ChirpStack deployment.

martinjaeger committed:
lorawan: add common backend for services

This is a prepartion for adding actual services needed for firmware
upgrade over the air (FUOTA).

The services run in a dedicated work queue.

This commit introduces code that initializes the work queue and
provides functions to schedule uplink messages after a given timeout.

Signed-off-by: Martin Jäger <martin@libre.solar>

martinjaeger committed:
lorawan: services: add Application Layer Clock Synchronization

This service allows to synchronize the clock with the application
server.

Synchronization requests are initiated by the device in a regular
interval, configurable via Kconfig.

The implementation only supports TS003-2.0.0, as the previous revision
TS003-1.0.0 requested to temporarily disable ADR and and set nb_trans
to 1. This causes issues on the server side and is not recommended
anymore.

Signed-off-by: Martin Jäger <martin@libre.solar>

JordanYates commented on subsys/lorawan/services/lorawan_services.c:
```diff
@@ -0,0 +1,181 @@
+		return -EFBIG;
+	}
+
+	timeout_abs_ticks = k_uptime_ticks() + timeout.ticks;
```
This doesn't work if the provided timeout is already absolute.

martinjaeger replied:
Fixed. I've replaced it with an `uint32_t timeout` in milliseconds instead of the opaque type to make it safer.

martinjaeger committed:
samples: subsys: lorawan: class_a: add documentation

Add a README for the LoRaWAN sample.

Signed-off-by: Martin Jäger <martin@libre.solar>

martinjaeger committed:
samples: subsys: lorawan: class_a: add clock sync option

Add an option to build the sample with application-layer clock sync
service running in the background.

This will also add a build-check for the clock sync service in CI.

Signed-off-by: Martin Jäger <martin@libre.solar>

martinjaeger commented:
Thanks a lot for your reviews @JordanYates @rerickson1 @Mani-Sadhasivam. I have addressed all your comments.

martinjaeger commented on subsys/lorawan/services/clock_sync.c:
```diff
@@ -0,0 +1,243 @@
+ */
+static int clock_sync_serialize_device_time(uint8_t *buf, size_t size)
+{
+	uint64_t device_time = k_uptime_get() / 1000 + ctx.time_correction;
```
Done. Using `-ENOSPC` now.

cfriedt replied:
I was referring more to the `1000`. It looks like it should be `MSEC_PER_SEC`.

cfriedt commented on subsys/lorawan/services/clock_sync.c:
```diff
@@ -0,0 +1,242 @@
+	__ASSERT(gps_time != NULL, "gps_time parameter is required");
+
+	if (ctx.synchronized) {
+		*gps_time = (uint32_t)(k_uptime_get() / 1000 + ctx.time_correction);
```
`MSEC_PER_SEC`

cfriedt commented on subsys/lorawan/services/clock_sync.c:
```diff
@@ -0,0 +1,242 @@
+ */
+static int clock_sync_serialize_device_time(uint8_t *buf, size_t size)
+{
+	uint64_t device_time = k_uptime_get() / 1000 + ctx.time_correction;
```
can this be `int32_t`?

cfriedt commented on subsys/lorawan/services/clock_sync.c:
```diff
@@ -0,0 +1,242 @@
+{
+	uint64_t device_time = k_uptime_get() / 1000 + ctx.time_correction;
+
+	if (size < 4) {
```
here, it would also be nicer to see `sizeof(device_time)`

cfriedt commented on subsys/lorawan/services/clock_sync.c:
```diff
@@ -0,0 +1,242 @@
+	buf[2] = (device_time >> 16) & 0xFF;
+	buf[3] = (device_time >> 24) & 0xFF;
+
+	return 4;
```
would be nicer to see `sizeof(device_time)` (if it can be `int32_t`) ... oh the joys of manual bitstuffing / serialization / deserialization

cfriedt reviewed (changes requested):
Just minor nits again with magic numbers

martinjaeger commented:
@Mani-Sadhasivam could you please have another look?

JordanYates commented on include/zephyr/lorawan/lorawan.h:
```diff
@@ -290,6 +290,38 @@ enum lorawan_datarate lorawan_get_min_datarate(void);
+ *
+ * This function uses the GPS epoch format, as used in all LoRaWAN services.
+ *
+ * The GPS epoch started on 1980-01-06T00:00:00Z, but has since diverged
```
It's probably worth being explicit here that the GPS epoch started on `1980-01-06T00:00:00Z UTC`, as opposed to any other epoch.

martinjaeger replied:
What do you mean by "being explicit"? The comment says exactly that, no?

JordanYates replied:
I agree that you can reconstruct the implication from the following sentence, but it is much clearer IMO to attach the epoch directly to the timestamp. You could reasonably infer that the epoch started on `1980-01-06T00:00:00Z TAI`, was divergent from `UTC` at the start, and has diverged further since. Timestamps and epochs are confusing enough without requiring the reader to make implications.

martinjaeger replied:
I still don't understand how you would like to improve the situation. Provide a conversion function to UTC or unix timestamps here?

I think that would be out of scope. The main purpose of the clock sync service right now is to make the other LoRaWAN services for FUOTA work. Conversion functions can be added later if desired (probably even at OS level and not as part of the LoRaWAN service).

JordanYates replied:
Sorry, all I am suggesting is to change the `1980-01-06T00:00:00Z` string to `1980-01-06T00:00:00Z UTC`, that's it.

martinjaeger replied:
Ahh, ok. But the `Z` already stands for UTC in [ISO 8601 timestamp format](https://en.wikipedia.org/wiki/ISO_8601#Coordinated_Universal_Time_(UTC)).

JordanYates commented on samples/subsys/lorawan/class_a/sample.yaml:
```diff
@@ -40,3 +40,8 @@ tests:
     extra_configs:
       - CONFIG_LORAMAC_REGION_RU864=y
+  sample.lorawan.class_a.clock_sync:
+    platform_allow: nucleo_wl55jc
```
`platform_allow` should be `integration_platforms`. From the docs:

```
platform_allow: <list of platforms>
Set of platforms that this test case should only be run for. Do not use this option to limit testing or building in CI due to time or resource constraints, this option should only be used if the test or sample can only be run on the allowed platform and nothing else.
```

JordanYates commented on subsys/lorawan/services/lorawan_services.h:
```diff
@@ -0,0 +1,57 @@
+ *
+ * @returns Services work queue handle
+ */
+struct k_work_q *lorawan_services_get_work_queue(void);
```
Does it make sense to have a `lorawan_services_work_reschedule(struct k_work_delayable *work, k_timeout_t timeout)`? This would save each of the services needing to query and store the queue handle locally.

martinjaeger commented:
@JordanYates and @Mani-Sadhasivam. Your requested changes have been implemented. Sorry for the delay, got stuck with other work in between...

JordanYates commented on subsys/lorawan/services/CMakeLists.txt:
```diff
@@ -0,0 +1,6 @@
+# SPDX-License-Identifier: Apache-2.0
+
+zephyr_library_sources_ifdef(CONFIG_LORAWAN_SERVICES
+        clock_sync.c
```
clock_sync.c should probably be under `_ifdef(CONFIG_LORAWAN_APP_CLOCK_SYNC`

martinjaeger replied:
True. Fixed.

JordanYates reviewed (dismissed):
Changes look good to me

martinjaeger commented:
@Mani-Sadhasivam can you please have a look again? The comments you had have been addressed.

Mani-Sadhasivam reviewed (approved):
LGTM! Thanks for the continuous push.
